### PR TITLE
Convert nodes.js to TypeScript

### PR DESCRIPTION
**Pull request by kevinb-khan**
## Summary:
After updating nodes.js to ues ES6 classes in #1747, the move to TypeScript became much more feasible.  While this PR mostly just adds types in a bunch of places, there are some changes to the code itself in order to minimize the number of casts and @ts-expect-errors that were require.  I've tried to keep these kinds of changes to a minimum.

I ended up converting a bunch of function expressions to arrow functions.  Hopefully that doesn't add too much additional overhead when reviewing.  There's still more cleanup that could be done (adding more missing types, converting all `var`s to either `const` or `let`, etc.), but I'm going to leave that for a future PR.

Issue: None

## Test plan:
- yarn test packages/kas/src

### CODE REVIEW

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [inline packages/kas/src/nodes.ts:2919]: I renamed this to `Sym` so that I could use `Symbol` for the prefixes types down below.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [inline packages/kas/src/nodes.ts:2924]: There were a few places that I made params `Readonly<>`.  This was an attempt to fix some issues with TS thinking that subclasses weren't assignable to a superclass.  This change probably isn't necessary so I'm going to try removing.  In general though, I think making things `Readonly<>` by default I preferable.  Maybe that could happening in a more comprehensive way in a future PR.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [inline packages/kas/src/nodes.ts:3086]: `vars` and `options` should probably be given real types.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [inline packages/kas/src/nodes.ts:132]: Instead of throwing in abstract methods, I think we can use TypeScript to mark the method as abstract, e.g.
```ts
abstract args(): (number | string | Expr)[];
```

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [inline packages/kas/src/nodes.ts:3623]: This can't be added to the `Num` class as static properties because `Int` and `Rational` depend on `Num` being defined.  We could probably drop the `Num` prefix.  Then we could export `Zero` and `One` directly here instead of down below.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts]: Even though we could do something other than using `Symbol` for the prefix types, I'd like avoid shadowing built-ins like this.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:76]: Not a blocker, just a thought: would it be worth having these as like static methods? Like `Expr.isInstance`? Then we could make it abstract and force children to implement it; that way each variation would have a builtin predicate? idk, maybe more trouble than its worth, I just noticed the `expr instanceof Mul` below which made me think of it.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:121]: I'm having a hard time deciphering this line. Would it be worth a comment explanation or something?

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:137]: I see a lot of `any`s, is that because we don't understand the types or because inheritance dictates what types will be used by different implementations?

**Review by handeyeco — COMMENTED** [review]
I'm going to have to circle back to this tomorrow. Just some initial thoughts for now.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:76]: I'm not sure if type predicates can be generic.  I'll have to experiment with that.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:121]: Will do.  The `new` part of the type let's TypeScript know that `func` should be a class.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:137]: This method is highly dynamic which makes typing it correctly difficult. 😞  I think we may be able to make this better using method overloads, but I don't want to block these changes on that.  I'll add a TODO here to tackle this later.  The return type though at least gives us some amount of type safety.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:214]: I think we could get rid of the reference to the regex now

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:139]: Could you talk a little about the switch from `isString` to `isString || isNumber`? I see that in a couple of places. For the most part we're just changing types, but this is changing logic right?

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:242]: Since `options` isn't being used, can it be omitted?

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:262]: This seems like a logic change, was it intentional? In the original one, if you passed `simplify({hello: "world"})` options would be `{once: false, hello: "world"}`; in the updated one it would be `{hello: "world"}`. I'm sure you checked, just wanted to be double sure.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:333]: Same here: do we need `options` if it's never used? Or do we need it in order to tell TS that this is implementing an abstract class or something?

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:351]: Is this essentially the same, just more explicit?

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:920]: **nit:** I kind of like having `hasDenom` as I think it makes intent clearer. Wouldn't lose sleep either way though.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:993]: `x instanceof Add` pops up a few times in the code. I wonder if it would be beneficial to make these predicates global and shared.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:1065]: Why "true"/"false" vs true/false?

EDIT: oh I see, `groupBy` is using the strings as keys in a return object. I think this is confusing, but your solution (`grouped.true`) is less confusing than the original (`grouped[true]`). I'm just wondering if we should have true/"true"/false/"false" as property names at all.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:1253]: I think this comment could use more context. Is there something we need to change? Something we need to look out for?

**Review by handeyeco — COMMENTED** [review]
Another batch of comments. Will continue looking at this after my meetings.

[ note to self, I left off at
```
1503                // expr.terms should always have at least two terms
1504                const last = _.last(expr.terms)!;
```
]

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:139]: Sometimes the args passed to node constructors can include numbers.  This is the case for `Num` and its subclasses.  I'm not sure why the existing code hasn't caused issues previously. 🤷‍♂️

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:242]: Yes it can.  I'll make this change.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:333]: We need it here because one of the subclasses uses it.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:351]: Without it TS thinks the return type is `any`.  Sometimes TS type inference is just really bad.  I tried turning on `noImplicitAny: true` in the tsconfig.json but then it was complaining about things that it was able to infer as a real type. 😞

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:993]: I'm a little hesitant to extract predicates into their own top-level function if they're only used in a single place.  I know there's a cost to redefining functions like this, but I don't believe that KAS performance is current an issue.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:1065]: I need to update this to use the `partition` helper function I added.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:1253]: I think this comment is stale.  I'll remove it.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:920]: I hear you.  TS unfortunately isn't able to follow the logic when it's expressed that way.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:2037]: These are hints from the parent class?

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:2183]: Do you mean an actual enum or a union? I was thinking we were generally trying to avoid enums.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:2446]: Good change 👍

File: packages/kas/src/nodes.ts
Comment (handeyeco) [reply packages/kas/src/nodes.ts:1000]: I'll defer to your judgement, but I'm a little confused by the "only used in a single place" part as `isAdd` is declared twice.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:2888]: Same here, just seems like going without destructuring is clearer and more concise.

```suggestion
        const termHasVar = hasVar(expr.terms[0])
        const a = termHasVar
            ? Mul.handleNegative(expr.terms[1])
            : Mul.handleNegative(expr.terms[0])
        const b = termHasVar
            ? Mul.handleDivide(expr.terms[0], variable)
            : Mul.handleDivide(expr.terms[1], variable)
```

But again, just personal preference.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:3129]: So this is just a placeholder because Num needs `n` for `getDecimalPlaces`? But really `n` gets set by its child classes? Just making sure I understand the motivation.

Also just curious, could this also be `n: number = 0;`?

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:3618]: Is this possible? I feel like probably not or else I imagine you would have done this.

```TS
    } catch (e: Error) {
        return {parsed: false, error: e.message;
    }
```

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:3748]: 🤨 This is some big brain JS.

I think `Symbol` makes sense. Also wonder if a string literal would work?

```
const hasPrefixes = "HAS_PREFIXES";
const hasntPrefixes = "HASNT_PREFIXES";

type Prefixes = hasPrefixes | hasntPrefixes;
```

A possible nice side effect is that console logging would make it visible whether it's has/hasnt

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:3834]: **nit against the original code:** I think using `map` here instead of `forEach` is semantically incorrect.

File: packages/kas/src/nodes.ts
Comment (handeyeco) [inline packages/kas/src/nodes.ts:3841]: Same here

**Review by handeyeco — APPROVED** [review]
Insomuch as I understand KAS, I think this PR makes sense. Thanks for doing all of this! I just ask that you don't land this right before going on sabbatical because I'm not sure any LEMS people understand this as much as you. 😂

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:1623]: That's correct.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:1949]: I don't think TS will refine `factor` to be a `Pow` correctly if we do that.  We'd have to repeat the `factor instanceof Pow` in each ternary.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:2037]: That's correct.  I believe that `Expr` sets `parens: false` or something like that.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:2183]: I can go with a union type.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:1000]: You're right.  I'll hoist it.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:2888]: I can do this one.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:3129]: This is to make TypeScript happy.  We could do `n: number = 0;`, but I tend to prefer setting variables in the constructor when there is a constructor.  I find it hard to reason about the order things run when we're doing both.  In this case I don't see any harm in having `n: number = 0;` outside the constructor.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts:3748]: Strings would be easier.  I'm not sure why the original author tried to be fancy with `hasPRefixes = {}`.  I'll switch these to strings.

File: packages/kas/src/nodes.ts
Comment (kevinb-khan) [reply packages/kas/src/nodes.ts]: I didn't even notice that.  Good eye.  There really should be a lint rule for useless `.map()` calls.  I ran into the same thing in webapp in a couple of places as well.

**Pull request merged**
